packages/glob/src/internal-globber.ts

```diff
@@ -0,0 +1,255 @@
+
+  private constructor() {}
+
+  getSearchPaths(): string[] {
+    // Return a copy
+    return this.searchPaths.slice()
+  }
+
+  async glob(): Promise<string[]> {
+    const result: string[] = []
+    for await (const itemPath of this.globGenerator()) {
+      result.push(itemPath)
+    }
+    return result
+  }
+
+  async *globGenerator(): AsyncGenerator<string, void> {
+    // Fill in defaults options
+    const options = globOptionsHelper.getOptions(this.options)
+
+    // Implicit descendants?
+    const patterns: Pattern[] = []
+    for (const pattern of this.patterns) {
+      patterns.push(pattern)
+      if (
+        options.implicitDescendants &&
+        (pattern.trailingSeparator ||
+          pattern.segments[pattern.segments.length - 1] !== '**')
+      ) {
+        patterns.push(
+          new Pattern(pattern.negate, pattern.segments.concat('**'))
+        )
+      }
+    }
+
+    // Push the search paths
+    const stack: SearchState[] = []
+    for (const searchPath of patternHelper.getSearchPaths(patterns)) {
+      core.debug(`Search path '${searchPath}'`)
+
+      // Exists?
+      try {
+        // Intentionally using lstat. Detection for broken symlink
+        // will be performed later (if following symlinks).
+        await fs.promises.lstat(searchPath)
+      } catch (err) {
+        if (err.code === 'ENOENT') {
+          continue
+        }
+        throw err
+      }
+
+      stack.unshift(new SearchState(searchPath, 1))
+    }
+
+    // Search
+    const traversalChain: string[] = [] // used to detect cycles
+    while (stack.length) {
+      // Pop
+      const item = stack.pop() as SearchState
+
+      // Match?
+      const match = patternHelper.match(patterns, item.path)
+      const partialMatch =
+        !!match || patternHelper.partialMatch(patterns, item.path)
+      if (!match && !partialMatch) {
+        continue
+      }
+
+      // Stat
+      const stats: fs.Stats | undefined = await DefaultGlobber.stat(
+        item,
+        options,
+        traversalChain
+      )
+
+      // Broken symlink, or symlink cycle detected, or no longer exists
+      if (!stats) {
+        continue
+      }
+
+      // Directory
+      if (stats.isDirectory()) {
+        // Matched
+        if (match & MatchKind.Directory) {
+          yield item.path
+        }
+        // Descend?
+        else if (!partialMatch) {
+          continue
+        }
+
+        // Push the child items in reverse
+        const childLevel = item.level + 1
+        const childItems = (await fs.promises.readdir(item.path)).map(
+          x => new SearchState(path.join(item.path, x), childLevel)
+        )
+        stack.push(...childItems.reverse())
+      }
+      // File
+      else if (match & MatchKind.File) {
+        yield item.path
+      }
+    }
+  }
+
+  /**
+   * Parses the input options and patterns to construct a new GlobberImpl instance.
+   */
+  static async parse(input: string): Promise<DefaultGlobber> {
```

Comment:
this parse function is new, everything else in this file basically moved into this class (was in glob.ts)